internal/daemon/api_files.go

```diff
@@ -0,0 +1,627 @@
+}
+
+func fileErrorToResult(err error) *errorResult {
+	if err == nil {
+		return nil
+	}
+	var kind errorKind
+	switch {
+	case errors.Is(err, os.ErrPermission):
+		kind = errorKindPermissionDenied
+	case errors.Is(err, os.ErrNotExist):
+		kind = errorKindNotFound
+	default:
+		kind = errorKindGenericFileError
+	}
+	return &errorResult{
+		Kind:    kind,
+		Message: err.Error(),
+	}
+}
+
+// Listing files
+
+func listErrorResponse(err error) Response {
+	status := http.StatusBadRequest
+	kind := errorKindGenericFileError
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		status = http.StatusNotFound
+		kind = errorKindNotFound
+	case errors.Is(err, os.ErrPermission):
+		status = http.StatusForbidden
+		kind = errorKindPermissionDenied
+	}
+	return &resp{
+		Type: ResponseTypeError,
+		Result: &errorResult{
+			Message: err.Error(),
+			Kind:    kind,
+		},
+		Status: status,
+	}
+}
+
+type fileInfoResult struct {
+	Path         string   `json:"path"`
+	Name         string   `json:"name"`
+	Type         fileType `json:"type"`
+	Size         *int64   `json:"size,omitempty"`
+	Permissions  string   `json:"permissions"`
+	LastModified string   `json:"last-modified"`
+}
+
+type fileType string
+
+const (
+	fileTypeFile      fileType = "file"
+	fileTypeDirectory fileType = "directory"
+	fileTypeSymlink   fileType = "symlink"
+	fileTypeSocket    fileType = "socket"
+	fileTypeNamedPipe fileType = "named-pipe"
+	fileTypeDevice    fileType = "device"
+	fileTypeUnknown   fileType = "unknown"
+)
+
+func fileModeToType(mode os.FileMode) fileType {
+	switch {
+	case mode&os.ModeType == 0:
+		return fileTypeFile
+	case mode&os.ModeDir != 0:
+		return fileTypeDirectory
+	case mode&os.ModeSymlink != 0:
+		return fileTypeSymlink
+	case mode&os.ModeSocket != 0:
+		return fileTypeSocket
+	case mode&os.ModeNamedPipe != 0:
+		return fileTypeNamedPipe
+	case mode&os.ModeDevice != 0:
+		return fileTypeDevice
+	default:
+		return fileTypeUnknown
+	}
+}
+
+func fileInfoToResult(fullPath string, info os.FileInfo) fileInfoResult {
+	mode := info.Mode()
+	var psize *int64
+	if mode.IsRegular() {
+		size := info.Size()
+		psize = &size
+	}
+	result := fileInfoResult{
+		Path:         fullPath,
+		Name:         path.Base(fullPath),
+		Type:         fileModeToType(mode),
+		Size:         psize,
+		Permissions:  fmt.Sprintf("%03o", mode.Perm()),
+		LastModified: info.ModTime().Format(time.RFC3339),
+	}
+	return result
+}
+
+func listFiles(pattern string, directoryItself bool) Response {
+	if !path.IsAbs(pattern) {
+		return statusBadRequest("pattern must be absolute (%q is not)", pattern)
+	}
+	result, err := listFilesErr(pattern, directoryItself)
+	if err != nil {
+		return listErrorResponse(err)
+	}
+	return SyncResponse(result)
+}
+
+func listFilesErr(pattern string, directoryItself bool) ([]fileInfoResult, error) {
+	info, err := os.Stat(pattern)
+	if errors.Is(err, os.ErrNotExist) {
+		dir, base := filepath.Split(pattern)
+		result, err := readDirFiltered(dir, base)
```

Comment:
I believe this is the same behavior as `ls`, though it's a little harder to see because of shell filename expansion:

```
$ touch fooa
$ touch foob
$ touch 'foo?'
$ ls foo?
'foo?'   fooa   foob
$ ls 'foo?'
'foo?'
```

This is only going to be an issue if you have filenames with `*` or `?` in them, right? In the spirit of being opinionated (cf our discussion about logging the other day), I'm okay just saying, "that's a really bad idea, don't do that".

Alternatively, if you feel strongly about this, we could make the list-files endpoint *always* take a pattern, or take either a `path` or `pattern` (or even `file`, `dir`, or `pattern` to make it even more explicit). If we go this route, my vote would be to only take `pattern` to keep it simple. Then you'd have to say `GET /v1/files?pattern=/etc/*` explicitly, which isn't necessarily a bad thing.

Reply:
This is not the same behavior as `ls`, since `ls` has no idea of wildcard expansion at all. The expansion is always done by the shell, and when you escape the string you disable the shell behavior altogether:
```
$ touch fooa foob 'foo?'
$ ls foo?
'foo?'   fooa   foob
$ ls 'foo?'
'foo?'
```
What I'm suggesting is that we do exactly that: we need two different modes, one in which wildcards are considered, and one where they are not, and we should not guess which is which based on the parameter alone, because that creates behavior people will be surprised and frustrated about.

Reply:
Fair enough. I've updated it to always have two parameters: `path` and `pattern`. Path is always a real file or directory path. Pattern is optional, and if specified, filters the list down to only the names which match the pattern. So the different ways you would call it are as follows:

```
path=/my/file               # get info about a single file (will still return list, but length 1)
path=/my/dir                # list info about all entries in directory
path=/my/dir&itself=true    # list info about directory itself
path=/my/dir&pattern=*.txt  # list info about names in directory that match *.txt
```

(I've also updated the name of the "directory" field to "itself" per your earlier comment.)